rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala

```diff
@@ -0,0 +1,453 @@
+}
+
+//A channel may be a quoted process, or a variable.
+sealed trait Channel
+
+  case class Quote(unquote: Proc[Channel]) extends Channel {
+    override def toString: String = "@(" + unquote + ")"
+  }
+
+  case class Var(id: String) extends Channel {
+    override def toString: String = id
+  }
+
+//A ChannelQueue may be an empty queue, a list of readers, or a list of writers.
+// It will never be both a list of readers and a list of writers
+sealed trait ChannelQueue
+
+  case class ReaderQueue(x:Reader, xs: List[Reader]) extends ChannelQueue {
+    override def toString: String = (x :: xs).map{_.toString}.mkString("[", "][", "]")
+  }
+
+  case class WriterQueue(x: Writer, xs: List[Writer]) extends ChannelQueue {
+    override def toString: String = (x :: xs).map{_.toString}.mkString("[", "][", "]")
+  }
+
+  case class EmptyQueue() extends ChannelQueue {
+    override def toString: String = "[]"
+  }
+
+//A reader is a abstraction providing a bound name "z", and a continuation
+//to evaluate, "k".
+sealed trait Reader
+
+  case class Abstraction(z: Channel, k: Proc[Channel]) extends Reader {
+    override def toString: String = " λ" + z + " { " + k.toString + " } "
+  }
+
+//A writer simply represents a message, "q".
+sealed trait Writer
+
+  case class Concretion(q: Channel) extends Writer {
+    override def toString: String = " " + q.toString + " "
+  }
+
+
+case class MachineState(store: Store, runQueue: RunQueue){
+  override def toString: String = "Queue: " + runQueue.mkString(" | ") + "\n" + "Store: " + store.mkString("{ ",", "," }") + "\n"
+}
+
+
+object RhoInterface {
+
+  val getStore: Trace[MachineState, Store] =
+    Trace.state { st => (st.store, st) }
+
+
+  val putStore: Store => Trace[MachineState, Unit] =
+    store =>
+      Trace.state { st0 => ((), MachineState(store, st0.runQueue)) }
+
+
+  val getRunQueue: Trace[MachineState, RunQueue] =
+    Trace.state { st => (st.runQueue, st) }
+
+
+  val putRunQueue: RunQueue => Trace[MachineState, Unit] =
+    runQ =>
+      Trace.state { st0 => ((), MachineState(st0.store, runQ)) }
+
+  // cancel(@*N) = N
+  val cancel: Channel => Channel = {
+    case Quote(Drop(n)) => n
+    case chan => chan
+  }
+
+  // Variable binding is represented by substitution in the body of the process.
+  // A more efficient implementation will achieve the same result using environments.
+  val bind: Channel => Channel => Proc[Channel] => Proc[Channel] =
+    chan0 =>
+      chan1 =>
+        proc =>
+          Proc.functorProc.map[Channel,Channel](proc) { name =>
+            //@*@Q = @Q in P{@*@Q/z}
+            if (name.equals(chan0)) cancel(chan1)
+            else name
+          }
+
+
+  val write: Channel => ChannelQueue => Trace[MachineState, Unit] =
+    chan =>
+      chanQ =>
+        for {store <- getStore
+             _ <- putStore(store + {
+               chan -> chanQ
+             })
+        } yield ()
```

Comment:
I don't think the `yield` keyword or brackets are necessary when you're not returning anything